tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -879,9 +879,8 @@ void CommandQueue::enqueue_read_buffer(Buffer& buffer, void* dst, bool blocking)
 
     if (buffer.buffer_layout() == TensorMemoryLayout::WIDTH_SHARDED or
         buffer.buffer_layout() == TensorMemoryLayout::BLOCK_SHARDED) {
-        convert_interleaved_to_sharded_on_host(dst, buffer.num_pages(),
-                                        buffer.page_size(),
-                                        buffer.dev_page_to_host_page_mapping(),
+        convert_interleaved_to_sharded_on_host(dst,
```

Comment:
isn't this an expensive operation considering that this code may be on critical path?

Reply:
this is an expensive operation but currently no other option. For a sharded tensor it's stored in a strided pattern, but for fast dispatch it needs to be contiguous in huge pages.
For height sharding the strided pattern ends up the same as row-major (default way of storing multi paged data)

The onus is on the developer. The developer has the option for width-sharded or block sharded tensors (not height sharded), to write in as interleaved buffer and use the interleaved_to_sharded op. This has a cost of an extra op on device but avoids doing the swap on host.